grpcbp/client_middlewares.go

```diff
@@ -101,3 +106,191 @@ func ForwardEdgeContextStreaming(ecImpl ecinterface.Interface) grpc.StreamClient
+		m := methodSlug(method)
+		activeRequestLabels := prometheus.Labels{
+			localServiceLabel:      serviceSlug,
+			methodLabel:            m,
+			remoteServiceSlugLabel: serverSlug,
+		}
+		clientActiveRequests.With(activeRequestLabels).Inc()
+
+		defer func() {
+			success := strconv.FormatBool(err == nil)
+			status, _ := status.FromError(err)
+
+			latencyLabels := prometheus.Labels{
+				localServiceLabel:      serviceSlug,
+				methodLabel:            m,
+				typeLabel:              unary,
+				successLabel:           success,
+				remoteServiceSlugLabel: serverSlug,
+			}
+
+			clientLatencyDistribution.With(latencyLabels).Observe(time.Since(start).Seconds())
+
+			rpcCountLabels := prometheus.Labels{
+				localServiceLabel:      serviceSlug,
+				methodLabel:            m,
+				typeLabel:              unary,
+				successLabel:           success,
+				remoteServiceSlugLabel: serverSlug,
+				codeLabel:              status.Code().String(),
+			}
+			clientRPCRequestCounter.With(rpcCountLabels).Inc()
+			clientActiveRequests.With(activeRequestLabels).Dec()
+		}()
+		return invoker(ctx, method, req, reply, cc, opts...)
+	}
+}
+
+// PrometheusStreamClientInterceptor is a client-side interceptor that provides Prometheus
+// monitoring for Streaming RPCs.
+//
+// It emits the following metrics:
+//
+// * grpc_client_active_requests gauge with labels:
+//
+//   - grpc_service: the local service slug, serviceSlug arg
+//   - grpc_method: the name of the method called on the gRPC service
+//   - grpc_slug: the name of the remote server the client connects to
+//
+// * grpc_client_latency_seconds histogram and grpc_client_requests_total
+//   counter with labels:
+//
+//   - grpc_service, grpc_method, grpc_slug
+//   - grpc_success: "true" if status is OK, "false" otherwise
+//   - grpc_type: type of request, i.e unary
+//   - grpc_code: the human-readable status code, e.g. OK, Internal, etc
+func PrometheusStreamClientInterceptor(serviceSlug, serverSlug string) grpc.StreamClientInterceptor {
+	return func(
+		ctx context.Context,
+		desc *grpc.StreamDesc,
+		cc *grpc.ClientConn,
+		method string,
+		streamer grpc.Streamer,
+		opts ...grpc.CallOption,
+	) (_ grpc.ClientStream, err error) {
+
+		clientStream, err := streamer(ctx, desc, cc, method, opts...)
+		if err != nil {
+			return nil, err
+		}
+		monitoredStream := newMonitoredClientStream(
+			clientStream,
+			method,
+			serviceSlug,
+			serverSlug,
+		)
+		return monitoredStream, nil
+	}
+}
+
+// monitoredClientStream wraps grpc.ClientStream to track Prometheus metrics when streaming starts and finishes.
+type monitoredClientStream struct {
+	grpc.ClientStream
+	start       time.Time
+	method      string
+	serviceSlug string
+	serverSlug  string
+}
+
+func newMonitoredClientStream(stream grpc.ClientStream, method, serviceSlug, serverSlug string) *monitoredClientStream {
+	s := &monitoredClientStream{
+		stream,
+		time.Now(),
+		method,
+		serviceSlug,
+		serverSlug,
+	}
```

Comment:
style nit: use keyed init for readability:
```suggestion
	s := &monitoredClientStream{
		ClientStream: stream,

		start: time.Now(),
		method: method,
		serviceSlug: serviceSlug,
		serverSlug: serverSlug,
	}
```

(🔕 also I still find the names `serviceSlug` vs. `serverSlug` confusing :( )

Reply:
will fix serviceSlug vs. serverSlug